Changing the Default retry value

vyasgun opened the pull request:
## Description

<!-- Please add a short summary about what the pull request is going to bring on the table -->
Modifying default retry value for backoff during resource updates
## Changes

<!-- Please add list of more detailed changes. These changes should be reflected also in the commit messages -->

* Changing the factor to 1.5 
* Changing MaxUpdateRetries to 5


## Reference

<!-- Please add the corresponding issue number which this pull request is about to fix -->
Fixes #

<!--
Please add an entry to CHANGELOG.adoc file, too, as part of your Pull Request.

In the following cases, add a short description of PR to the unreleased section in CHANGELOG.adoc:

- 🎁 New feature
- 🐛 Bug fix
- ✨ Feature Update
- 🐣 Refactoring
- 🗑️ Remove feature or internal logic

See other entries in CHANGELOG.adoc as an example for how to add the entry, including a reference to the corresponding issue/PR

PLEASE DON'T ADD THAT LINE HERE IN THE PULL-REQUEST DESCRIPTION BUT DIRECTLY IN CHANGELOG.ADOC AND ADD CHANGELOG.ADOC AS PART OF YOUR PULL-REQUEST.
-->

<!--
To automatically lint go code in this pull request uncomment the line below. You get feedback as comments on your pull request then -->

<!--
/lint
-->

rhuss commented on pkg/util/compare.go:
```diff
@@ -17,10 +17,25 @@ package util
+	// DefaultRetry is the default set of rules
+	// to be followed when retrying for conflicts during
+	// resource update
+	DefaultRetry = wait.Backoff{
```
I would not put the default to that (unrelated) util package but move it more closely to the API where it is used. I know, it is used in several different clients where this is used, so finding a proper place is tricky. But this package is for sure not the correct place ;-) For example, if we would have an intermediate package "clients" in `pkg/clients` where all our clients to the backend are collected, then this would be the proper package (we don't have that yet, but maybe that would be a good idea ?)

For now, I would not hesitate to introduce a `pkg/config` with a single file `defaults.go` in a package "config". Might sound like overkill at first, but will help future refactoring if we clearly separate the concerns.

(my general rule of thumb: If one does not know where to put a shared constant, a global `util` package is usually a bad idea (and `util` should be the last choice considered anyway in any case :). Utilities should be independent of a concrete domain and generally reusable for a wide range of scenarios)

vyasgun replied:
I wasn't sure about the file to put in. I have moved the variable to the suggested location. I put it in util because the DefaultRetry variable is client-go package is in k8s.io/client-go/util/retry/util.go so why is it incorrect in our case?

rhuss reviewed (commented):
Thanks a lot for the quick turnaround.

Looks good, but see my comment where to place the default. Also, I think we could remove "MaxUpdateRetries" everywhere in favor of setting this in the default directly. This would also reduce the footprint of the method signatures as we don't have to carry over the retry (which typically is not changed anyway (alternatively change the `nrRetries` parameter to use a full backoff. Maybe make the backoff an optional parameter (via `...`) ?

maximilien reviewed (commented):
Thanks for contribution, however, can you also please add a bit more info (rationale) why these new values are better. Perhaps you have tests that shows increasing from 3 to 5 is better. Otherwise, looks like we are just guessing...

rhuss commented:
5 is the default value chosen by Kubernetes --> xhttps://github.com/knative/client/blob/ad4a7b18a2e8724d9ca08311a64f28a0adb73c47/vendor/k8s.io/client-go/util/retry/util.go#L26-L33

So the question is more, why we haven chosen 3 ;-) Actually the reason we area fiddling around with the values is that we still see CI flakes with conflict errors.

vyasgun commented:
@rhuss @maximilien I have added a comment for the variable to specify that all fields except `duration` and `factor` have the same value as the client-go's `DefaultRetry`.

vyasgun commented:
@rhuss If we refactor the retry functions to accept optional number of nrRetries, we will have to add a validation step for not accepting more than 1 arg. Would it be better to just change the signature so it doesn't accept any nrRetries at all?

dsimansk commented on pkg/kn/commands/source/ping/update.go:
```diff
@@ -93,7 +94,7 @@ func NewPingUpdateCommand(p *commands.KnParams) *cobra.Command {
 			}
-			backoff := retry.DefaultRetry
+			backoff := config.DefaultRetry
 			backoff.Steps = MaxUpdateRetries
```
Do we need to actually set the `.Steps` on top of default value, since `MaxUpdateRetries` is changed to `5` everywhere.

vyasgun commented:
@dsimansk Thanks for the review. I have changed the argument to `DefaultRetry.Steps` for now and removed `MaxUpdateRetries` constants from all packages.

vyasgun commented:
/retest

dsimansk commented on pkg/kn/commands/service/create.go:
```diff
@@ -208,7 +209,7 @@ func prepareAndUpdateService(ctx context.Context, client clientservingv1.KnServi
 		if err != nil {
 			// Retry to update when a resource version conflict exists
-			if apierrors.IsConflict(err) && retries < MaxUpdateRetries {
+			if apierrors.IsConflict(err) && retries < config.DefaultRetry.Steps {
```
I know this's only changed due to removal of variable, but looking at the code itself I wonder if shouldn't be just `.UpdateServiceWithRetry` without that additional conflict handling block.

dsimansk commented on pkg/kn/commands/source/ping/ping.go:
```diff
@@ -25,7 +25,7 @@ import (
 
 const (
-	MaxUpdateRetries = 3
+	MaxUpdateRetries = 5
```
@vyasgun one more leftover const, otherwise it's OK.

dsimansk commented on pkg/kn/commands/trigger/trigger.go:
```diff
@@ -22,7 +22,7 @@ import (
 const (
 	// How often to retry in case of an optimistic lock error when replacing a trigger (--force)
-	MaxUpdateRetries = 3
+	MaxUpdateRetries = 5
```
Same here.

dsimansk commented on pkg/kn/commands/source/ping/ping.go:
```diff
@@ -25,7 +25,7 @@ import (
 
 const (
-	MaxUpdateRetries = 3
+	MaxUpdateRetries = 5
```
@vyasgun one more leftover const, otherwise it's OK.

vyasgun replied:
Thanks for pointing out. Fixed the remaining consts

dsimansk commented:
Thanks @vyasgun!

/approve
/lgtm

Adding hold for @rhuss @maximilien to take a look.
/hold

rhuss commented on pkg/eventing/v1/client.go:
```diff
@@ -148,7 +150,7 @@ func (c *knEventingClient) UpdateTriggerWithRetry(ctx context.Context, name stri
 
 func updateTriggerWithRetry(ctx context.Context, c KnEventingClient, name string, updateFunc TriggerUpdateFunc, nrRetries int) error {
-	b := retry.DefaultRetry
+	b := config.DefaultRetry
```
Looks good to me, except that I probably would change the signature/API instead of taking a retry number to take a full backoff struct for more flexibility.

But happy to merge now, too, we could adapt later as well.

vyasgun replied:
Since the UpdateServiceWithRetry function is being used in the func plugin as well, I think we can change the signature later?

dsimansk commented:
After chat with @rhuss, it should be good to go. With a minor future improvement stated in the comment.
/unhold
